R/BackendHdf5-class.R

```diff
@@ -0,0 +1,149 @@
+#' to the @hdf5file slot.
+#'
+#' @param object `BackendHdf5`
+#'
+#' @param files `character` with the original file names.
+#'
+#' @param spectraData `DataFrame` with the spectrum metadata.
+#'
+#' @param path `character(1)` defining the path where the hdf5 files should be
+#'     stored.
+#'
+#' @return `BackendHdf5` object.
+#'
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @noRd
+setMethod("backendInitialize", "BackendHdf5", function(object, files,
+                                                       spectraData, path = ".",
+                                                       ...) {
+    path <- suppressWarnings(normalizePath(path))
+    dir.create(path, showWarnings = FALSE, recursive = TRUE)
+    n_files <- length(files)
+    object@files <- files
+    object@md5sum <- character(n_files)
+    object@hdf5file <- character(n_files)
+    comp_level <- .hdf5_compression_level()
+    for (i in seq_len(n_files)) {
+        h5file <- paste0(path, "/", digest(files[i]), ".h5")
+        if (file.exists(h5file))
+            stop("File ", h5file, " does already exist. Please use a different path.")
+        h5 <- H5Fcreate(h5file)
+        h5createGroup(h5, "spectra")
+        h5createGroup(h5, "md5")
+        H5Fclose(h5)
+        object@hdf5file[i] <- h5file
+    }
+    validObject(object)
+    object
+})
+
+#' Import the data from the raw MS files and store them to the hdf5 files.
+#'
+#' @inheritParams backendInitialize
+#'
+#' @return `BackendHdf5`
+#'
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @noRd
+setMethod("backendImportData", "BackendHdf5", function(object, spectraData,
+                                                       ...,
+                                                       BPPARAM = bpparam()) {
+    object@md5sum <- bpmapply(fileNames(object), object@hdf5file,
+                              FUN = .serialize_msfile_to_hdf5,
+                              BPPARAM = BPPARAM)
+    validObject(object)
+    object
+})
+
+## setMethod("backendReadSpectra", "BackendHdf5", function(object, spectraData,
+##                                                         ...) {
+## })
+
+## setMethod("backendWriteSpectra", "BackendHdf5", function(object, spectra,
+##                                                          spectraData) {
+## })
+
+#' Write the content of a single mzML/etc file to an h5file. We're using the
+#' spectrum index in the file as data set ID.
+#'
+#' @return `character(1)` with the md5 sum of the spectra data.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.serialize_msfile_to_hdf5 <- function(file, h5file) {
+    h5 <- H5Fopen(h5file)
+    comp_level <- .hdf5_compression_level()
+    fh <- openMSfile(file)
+    hdr <- header(fh)
+    pks <- peaks(fh)
+    close(fh)
+    for (i in seq_along(pks)) {
+        .pks <- pks[[i]]
+        colnames(.pks) <- c("mz", "intensity")
+        h5write(.pks, h5, paste0("spectra/", i), level = comp_level)
```

Comment:
I use the `H5Gopen` in my prototype to avoid creating this `group/id` string every time: https://gist.github.com/sgibb/54c1477d26fa8702231c447badd11776#file-class-draft-r-L198
But I am not sure that this is useful at all.

Reply:
Could you do some performance tests on that?

Reply:
If we put the `paste0("spectra/", i)` out of the loop it makes no difference. I would suggest my slightly adapted version:
```r
.mzr2hdf5 <- function(file, h5) {
    msfh <- openMSfile(file)
    ## read complete header first to avoid random seq fault errors from
    ## proteowizard
    hdr <- header(msfh)
    on.exit(close(msfh))

    lvl <- .hdf5_compression_level()

    h5fh <- H5Fopen(h5)
    on.exit(H5Fclose(h5fh), add=TRUE)

    nms <- paste0("spectra/", seq_along(msfh))

    for (i in seq(along=msfh)) {
        h5write(peaks(msfh, i), h5fh, nms[i], level=lvl)
    }
}

Reply:
totally agree. I did that already in an updated version.